Add support for OAuth2.0 server-to-server authentication

dimitarproynov opened the pull request:
Added two new provider config properties: ClientID and ClientSecret that represent the credentials in OAuth2.0 client_credentials flow (server-to-server) and are in conflict with APIToken.

Implementation of OAuth2.0 App based authentication, based on prior work:
https://github.com/vmware/terraform-provider-vmc/pull/32

VMware, own docs:
https://developer.vmware.com/apis/csp/csp-iam/latest/csp/gateway/am/api/auth/token/post/

The change includes a major refactor to encapsulate all authentication logic into the connector.Wrapper struct, as different parts of authentication logic was leaking all over the place. Now the only place in the code where details about API_TOKEN or ClientID/ClientSecret are the Authenticate() method of connector.Wrapper.

Also converted Acceptance tests to run with ClientID and ClientSecret instead of my own APIToken. Since the test OAuth App runs with minimal priviledges some nil checks were added to reflect the API not returning some sensitive data (even if it is Zerocloud dummy data).

This is a new feature requested by customers and specifically StateStreet:
https://github.com/vmware/terraform-provider-vmc/issues/164

Testing done:
make build
make test

golangci-lint run

Full Acceptance tests successful run:
https://github.com/dimitarproynov/terraform-provider-vmc/actions/runs/4103883503/jobs/7078645332

Commit:
Add support for OAuth2.0 server-to-server authentication

Testing done:
make build
make test
golangci-lint run

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

Commit:
Merge remote-tracking branch 'origin/main' into OAuth_app_authn

Commit:
Pass E2E tests with OAuth App ClientID/ClientSecret instead of ApiToken

Some nil checks were added as the OAuth App has minimum set of roles.

Testing done:
make build
make test
golangci-lint run

 make testacc TESTARGS="-run='TestAccResourceVmcSddcZerocloud|TestAccResourceVmcClusterZerocloud|TestAccResourceVmcSiteRecoveryZerocloud|TestAccResourceVmcSrmNodeZerocloud|TestAccDataSourceVmcCustomerSubnetsBasic|TestAccDataSourceVmcConnectedAccountsBasic|TestAccDataSourceVmcOrgBasic|TestAccDataSourceVmcSddcBasic|TestAccResourceSddcGroupZerocloud' -parallel 4"
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run='TestAccResourceVmcSddcZerocloud|TestAccResourceVmcClusterZerocloud|TestAccResourceVmcSiteRecoveryZerocloud|TestAccResourceVmcSrmNodeZerocloud|TestAccDataSourceVmcCustomerSubnetsBasic|TestAccDataSourceVmcConnectedAccountsBasic|TestAccDataSourceVmcOrgBasic|TestAccDataSourceVmcSddcBasic|TestAccResourceSddcGroupZerocloud' -parallel 4 -timeout 240m
?   	github.com/vmware/terraform-provider-vmc	[no test files]
=== RUN   TestAccDataSourceVmcConnectedAccountsBasic
--- PASS: TestAccDataSourceVmcConnectedAccountsBasic (4.84s)
=== RUN   TestAccDataSourceVmcCustomerSubnetsBasic
=== PAUSE TestAccDataSourceVmcCustomerSubnetsBasic
=== RUN   TestAccDataSourceVmcOrgBasic
--- PASS: TestAccDataSourceVmcOrgBasic (4.28s)
=== RUN   TestAccDataSourceVmcSddcBasic
--- PASS: TestAccDataSourceVmcSddcBasic (5.28s)
=== RUN   TestAccResourceVmcClusterZerocloud
=== PAUSE TestAccResourceVmcClusterZerocloud
=== RUN   TestAccResourceSddcGroupZerocloud
=== PAUSE TestAccResourceSddcGroupZerocloud
=== RUN   TestAccResourceVmcSddcZerocloud
=== PAUSE TestAccResourceVmcSddcZerocloud
=== RUN   TestAccResourceVmcSiteRecoveryZerocloud
=== PAUSE TestAccResourceVmcSiteRecoveryZerocloud
=== RUN   TestAccResourceVmcSrmNodeZerocloud
=== PAUSE TestAccResourceVmcSrmNodeZerocloud
=== CONT  TestAccDataSourceVmcCustomerSubnetsBasic
=== CONT  TestAccResourceVmcSddcZerocloud
=== CONT  TestAccResourceVmcSrmNodeZerocloud
=== CONT  TestAccResourceSddcGroupZerocloud
--- PASS: TestAccDataSourceVmcCustomerSubnetsBasic (11.24s)
=== CONT  TestAccResourceVmcSiteRecoveryZerocloud
SDDC terraform_sddc_test_ty6ec8rg84 created successfully with id d5b426e1-af69-48f5-9060-9c3e60012246
--- PASS: TestAccResourceVmcSddcZerocloud (120.73s)
=== CONT  TestAccResourceVmcClusterZerocloud
--- PASS: TestAccResourceVmcSiteRecoveryZerocloud (190.80s)
--- PASS: TestAccResourceVmcSrmNodeZerocloud (236.71s)
--- PASS: TestAccResourceSddcGroupZerocloud (246.43s)
Cluster for SDDC d2495363-f0af-44a6-9491-59429a63a0d9 created successfully
--- PASS: TestAccResourceVmcClusterZerocloud (480.60s)
PASS
ok  	github.com/vmware/terraform-provider-vmc/vmc	615.960s
?   	github.com/vmware/terraform-provider-vmc/vmc/connector	[no test files]
?   	github.com/vmware/terraform-provider-vmc/vmc/constants	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/vmware/terraform-provider-vmc/vmc/sddcgroup	(cached) [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/vmware/terraform-provider-vmc/vmc/task	(cached) [no tests to run]

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

dimitarproynov committed:
Merge pull request #11 from dimitarproynov/OAuth_app_authn

OAuth app authn impl merge into main

karaatanassov commented on vmc/connector/clientconnector.go:
```diff
@@ -26,23 +27,46 @@ type Authenticator interface {
 	CspURL       string
 }
 
+func CopyWrapper(original Wrapper) *Wrapper {
```
Possibly dumb question. 

Why not use the copy operator like

`newWrapper Wrapper = *wrapperPtr`

The way you declared it just returning pointer to `original` will do the trick I think. `original` is a copy.

dimitarproynov replied:
Not dumb at all. You're correct. Addressed.

karaatanassov commented on vmc/data_source_vmc_sddc.go:
```diff
@@ -212,9 +212,14 @@ func dataSourceVmcSddcRead(d *schema.ResourceData, m interface{}) error {
-			d.Set("nsxt_cloudadmin_password", *sddc.ResourceConfig.NsxCloudAdminPassword)
+			// Evade nil pointer dereference when user's access_token doesn't have NSX roles
+			if sddc.ResourceConfig.NsxCloudAdminPassword != nil {
+				_ = d.Set("nsxt_cloudadmin_password", *sddc.ResourceConfig.NsxCloudAdminPassword)
```
nit: do you need the `_ =`? I see line 212 does not have it.

dimitarproynov replied:
The IDE complains I'm not using the result ot d.Set and presents it as a warning. Line 212 doesn't have it as I want to limit the changes on the side. Possibly there will be a dedicated change to solve all the warnings.

karaatanassov reviewed (approved):
There are couple minor comments you may want to look at.

fstrudel commented on vmc/connector/clientconnector.go:
```diff
@@ -82,15 +106,74 @@ func SecurityContextByRefreshToken(refreshToken string, cspURL string) (core.Sec
+	return connector, nil
+}
+
+func securityContextByClientID(clientID string, clientSecret string, cspURL string) (core.SecurityContext, error) {
```
Is there a reason not to use some standard packages to perform the OAuth2 protocol? like https://pkg.go.dev/golang.org/x/oauth2/clientcredentials ?

dimitarproynov replied:
No reason. I didn't know they existed. Reworked the mentioned fuction with the clientcredentials package.

fstrudel commented on vmc/constants/constants.go:
```diff
@@ -10,9 +10,12 @@ const (
 	CspRefreshURLSuffix string = "/csp/gateway/am/api/auth/api-tokens/authorize"
 
+	// CspOauthURLSuffix defines the CSP Oauth API endpoint.
+	CspOauthURLSuffix string = "/csp/gateway/am/api/auth/token"
```
It's called the "token endpoint", so you should call it `CspTokenURL` or something like that (even though "CSP" is not a user facing term, "VMware cloud services" is)

dimitarproynov replied:
Addressed.

Commit:
Address issues pointed out by reviewers.

Use the golang.org/x/oauth2/clientcredentials library instead of manually crafting a Http request to CSP for OAuth2.0 authentication flow.

Testing done:
make build
make test
golangci-lint run

terraform-provider-vmc % make testacc TESTARGS="-run='TestAccResourceVmcSddcZerocloud'"
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run='TestAccResourceVmcSddcZerocloud' -timeout 240m
?   	github.com/vmware/terraform-provider-vmc	[no test files]
=== RUN   TestAccResourceVmcSddcZerocloud
=== PAUSE TestAccResourceVmcSddcZerocloud
=== CONT  TestAccResourceVmcSddcZerocloud
SDDC terraform_sddc_test_eucsf7qmk2 created successfully with id eb298641-625d-491b-8954-ea5311158965
--- PASS: TestAccResourceVmcSddcZerocloud (122.66s)
PASS
ok  	github.com/vmware/terraform-provider-vmc/vmc	122.930s
?   	github.com/vmware/terraform-provider-vmc/vmc/connector	[no test files]
?   	github.com/vmware/terraform-provider-vmc/vmc/constants	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/vmware/terraform-provider-vmc/vmc/sddcgroup	0.200s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/vmware/terraform-provider-vmc/vmc/task	0.184s [no tests to run]

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>
